map.go

```diff
@@ -1622,12 +1649,56 @@ func newMapIterator(target *Map) *MapIterator {
 	}
 }
 
+// cursorToKeyOut copies the current value held in the cursor to the
+// provided argument. In case of errors, returns false and sets a
+// non-nil error in the MapIterator.
+func (mi *MapIterator) cursorToKeyOut(keyOut interface{}) bool {
+	buf := mi.cursor.([]byte)
+	if ptr, ok := keyOut.(unsafe.Pointer); ok {
+		copy(unsafe.Slice((*byte)(ptr), len(buf)), buf)
+	} else {
+		mi.err = sysenc.Unmarshal(keyOut, buf)
+	}
+	return mi.err == nil
+}
+
+// fetchNextKey loads into the cursor the key following the provided one.
+func (mi *MapIterator) fetchNextKey(key interface{}) bool {
+	mi.err = mi.target.NextKey(key, mi.cursor)
+	if mi.err == nil {
+		return true
+	}
+
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		mi.done = true
+		mi.err = nil
+	} else {
+		mi.err = fmt.Errorf("get next key: %w", mi.err)
+	}
+
+	return false
+}
+
+// drainMapEntry removes and returns the key held in the cursor
+// from the underlying map.
+func (mi *MapIterator) drainMapEntry(valueOut interface{}) bool {
+	mi.err = mi.target.LookupAndDelete(mi.cursor, valueOut)
+	if mi.err == nil {
+		mi.count++
+		return true
+	}
+
+	if errors.Is(mi.err, ErrKeyNotExist) {
+		mi.err = nil
+	} else {
```

Comment:
Nitpicking, but IMO this is important none the less: try to avoid `else` near the end of a block, it's often unnecessary.

In this case, it's also a little misleading when mixed in with early returns. The `else` clause would be executed if `!errors.Is(..., ErrNotExist)`, but also when `err == nil`, and the latter case is already checked above.

It's usually clearer to be explicit and let things fall through naturally:
```go
	if errors.Is(mi.err, ErrKeyNotExist) {
		mi.err = nil
		return false
	}

	mi.err = fmt.Errorf("lookup and delete next key: %w", mi.err)
	return false
```